File: sample1/python/writer.py
""" 

Protocol buffer writer in Python

author: Atsushi Sakai

"""

import addressbook_pb2


def main():
    print("start!!")

    address_book = addressbook_pb2.AddressBook()

    person1 = address_book.people.add()
    person1.id = 1234
    person1.name = "John Doe"
    person1.email = "jdoe@example.com"
    phone = person1.phones.add()
    phone.number = "555-4321"
    phone.type = addressbook_pb2.Person.HOME

    # person1.no_such_field = 1  # raises AttributeError
    # person1.id = "1234"        # raises TypeError

    person2 = address_book.people.add()
    person2.id = 4321
    person2.name = "Tom Ranger"
    person2.email = "tranger@example.com"
    phone = person2.phones.add()
    phone.number = "555-4322"
    phone.type = addressbook_pb2.Person.WORK

    print(address_book)  # Human readerble print

    # writing the data
    f = open("pbdata.dat", "wb")
    f.write(address_book.SerializeToString())
    f.close()

    print("done!!")


if __name__ == '__main__':
    main()

File: sample1/julia/writer.jl
"
    Protocol buffer data writer in Julia 

    author: Atsushi Sakai
"

using ProtoBuf

include("tutorial.jl")
include("addressbook_pb.jl")

function main()
    println(PROGRAM_FILE," start!!")

    address_book = AddressBook(
                               people=Person[]
                              )

    person1 = Person(
                     id = 1234,
                     name = "John Doe",
                     email = "jdoe@example.com",
                     phones = Person_PhoneNumber[]
                    )

    phone = Person_PhoneNumber(
                               number="555-4321",
                               _type = __enum_Person_PhoneType().HOME
                              )
    push!(person1.phones, phone)

    push!(address_book.people, person1)
 
    
    person2 = Person(
                     id = 4321,
                     name = "Tom Ranger",
                     email = "tranger@example.com",
                     phones = Person_PhoneNumber[]
                    )

    phone = Person_PhoneNumber(
                               number="555-4322",
                               _type = __enum_Person_PhoneType().WORK
                              )
    push!(person2.phones, phone)

    push!(address_book.people, person2)
 

    println(address_book)  # Human readable print

    iob = PipeBuffer();
    writeproto(iob, address_book)

    open( "pddata_julia.dat", "w" ) do fp
        write( fp, iob )
    end

    println(PROGRAM_FILE," Done!!")
end


main()

